File: leetcode/0048-RotateImage/solution.go
// compute Complexity O(N*N)
// Space Complexity O(1)

// Test cases
// []
// [1]
// [a b, c d] i:{0} j:{0}
// [a b c, d e f, g h i]

func rotate(matrix [][]int)  {
    n := len(matrix)
    for i := 0; i < n / 2; i++  {
        for j := i; j < (n -1 -i); j++ {
            tmp := matrix[i][j]
            matrix[i][j] = matrix[n-1-j][i]
            matrix[n-1-j][i] = matrix[n-1-i][n-1-j]
            matrix[n-1-i][n-1-j] = matrix[j][n-1-i]
            matrix[j][n-1-i] = tmp;
        }
    }
}

File: leetcode/0048-RotateImage/solution.cpp
// Compute Complexity O(N*N)
// space complexity O(1)

class Solution {
    public:
        void rotate(vector<vector<int>>& matrix) {
            const auto n = matrix.size();
            for (int i = 0; i <= (n-1)/2; i++) {
                for (int j = i; j < (n - 1 - i); j++) {
                    std::cout << i << " " << std::endl;
                    auto tmp = matrix[i][j];
                    matrix[i][j] = matrix[n-j-1][i];
                    matrix[n-j-1][i] = matrix[n-i-1][n-j-1];
                    matrix[n-i-1][n-j-1] = matrix[j][n-1-i];
                    matrix[j][n-1-i] = tmp;
                }
            }
        }
    };
    